### PR TITLE
NIFI-8927 - Add option to start/stop all controllers

### PR DESCRIPTION
**Pull request by eduardofontes**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-8927._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
From a pure testing perspective, this looks good. Not sure if there are corner cases though... but in the end you're just calling exposed REST APIs so I don't see why there would be an issue. This is definitely helpful. I'd still wait for someone familiar with the UI to have a look. Thanks @eduardofontes 

Screenshot 2021-07-27 at 16 30 15
Screenshot 2021-07-27 at 16 31 54

**Comment by markobean** [comment]
I suggest you update references to the phrasing "controllers" to "controller services". This is most important in the user-facing UI, but also will add clarity in the code as well.

**Comment by eduardofontes** [comment]
@markobean Done!

**Comment by eduardofontes** [comment]
Could somebody prease review this PR?

**Comment by joewitt** [comment]
We need to have more information on what has been verified here.  What happens on a taxed system with many controller services?  Does this cause blocking while that is happening?  How does this play with the async handling?  I'd be hesitant to merge until someone more knowledgeable has looked at this AND we have some information on testing.

Obviously this is a useful feature but needs engagement from more folks in the community with more knowledge on these parts.

**Comment by mcgilman** [comment]
@eduardofontes Thanks for the PR! The process of enabling and disabling Controller Services is complex. Currently, the NiFi UI supports enabling and disabling a single Controller Service. When enabling the user has the option to conditionally enable any dependent Controller Services and starting any referencing Processors/Reporting Tasks. When disabling the user does not have this option as the user must stop all referencing Processors/Reporting Tasks and disable any dependent Controller Services. Since there could be any number of affected components, this helps keep the operation as concise as possible, helps guarantee its success, and provides the user the necessary insight into the operation.

The API that your invoking is actually part of a larger sequence [1] to update Variables and Parameters that could be referenced by Controller Services and their referencing components across all Nodes in a cluster. The UX for those actions show the user the affected components and provides updates as the cluster performs the sequence of operations. I think this feature would also benefit from a similar UX where the user can see the full scope of their action and gets updates while the action completes.

[1] https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java#L827

**Comment by markap14** [comment]
Thanks for your thoughts @joewitt and @mcgilman. Here's my 2 cents:
- I'm not the one to review the code because my JavaScript skills are non-existent, so I can't review from that perspective.
- The calls that the UI makes are to the existing endpoints, which asynchronously enable/disable controller services. So I don't think we have concerns about high load causing timeouts, failures, etc. because it shouldn't block
- I do agree with mcgilman that I'd like to see a UI that allows the controller services to be visualized with all referencing components so that users know exactly what they are doing when they Disable All. And when enabling, ideally it would be helpful to be able to choose to start referencing components as well.
- That said, I think this feature is extremely useful as-is. The more powerful UI capability can still be provided in a future release of NiFi, but I think this is significantly more helpful than what we have currently.

So while I'm not in a position to review the code I'm a +1 personally.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 1]
```diff
@@ -845,6 +845,116 @@
             }
         },
 
+        /**
+         * Enable all controller services in the specified ProcessGroup.
+         *
+         * @argument {selection} selection      The selection
+         * @argument {cb} callback              The function to call when request is processed
+         */
+         enableAllControllerServices: function (selection,cb) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:854]: What is `cb` used for in this case? Other actions (like `stop` and `terminate`) had a callback introduced when we introduced the ability to stop a running Processor from the details dialog and subsequently open the configuration dialog (NIFI-5986). I don't see a similar capability here with enabling a Controller Service.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 2]
```diff
@@ -845,6 +845,116 @@
...
+            }
+        },
+
+        /**
+         * Disable all controller services in the specified ProcessGroup.
+         *
+         * @argument {selection} selection      The selection
+         * @argument {cb} callback              The function to call when request is processed
+         */
+         disableAllControllerServices: function (selection,cb) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:909]: Same comment as above regarding the `cb`.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 3]
```diff
@@ -845,6 +845,116 @@
...
+         enableAllControllerServices: function (selection,cb) {
+            if (selection.empty()) {
+                // build the entity
+                var entity = {
+                    'id': nfCanvasUtils.getGroupId(),
+                    'state': 'ENABLED'
+                };
+                updateResource(config.urls.api + '/flow/process-groups/' + encodeURIComponent(nfCanvasUtils.getGroupId()) + '/controller-services', entity).done(updateProcessGroup);
+            } else {
+                var componentsToStart = selection.filter(function (d) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:863]: The conditions on menu-items that could invoke this function guarantee that the selection is either empty or a single Process Group. Unless this is being invoked in a different context, we should update this to work with a single element.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 4]
```diff
@@ -845,6 +845,116 @@
...
+         disableAllControllerServices: function (selection,cb) {
+            if (selection.empty()) {
+                // build the entity
+                var entity = {
+                    'id': nfCanvasUtils.getGroupId(),
+                    'state': 'DISABLED'
+                };
+                updateResource(config.urls.api + '/flow/process-groups/' + encodeURIComponent(nfCanvasUtils.getGroupId()) + '/controller-services', entity).done(updateProcessGroup);
+            } else {
+                var componentsToStop = selection.filter(function (d) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:918]: Same comment as above regarding the selection.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 5]
```diff
@@ -845,6 +845,116 @@
...
+                    'state': 'ENABLED'
+                };
+                updateResource(config.urls.api + '/flow/process-groups/' + encodeURIComponent(nfCanvasUtils.getGroupId()) + '/controller-services', entity).done(updateProcessGroup);
+            } else {
+                var componentsToStart = selection.filter(function (d) {
+                    return nfCanvasUtils.isProcessGroup(d3.select(this));
+                });
+
+                // ensure there are some component to start
+                if (!componentsToStart.empty()) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:868]: Controller Services are not started/stopped. They are enabled/disabled. Can we update the names/comments to reflect this?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 6]
```diff
@@ -845,6 +845,116 @@
...
+                    'state': 'DISABLED'
+                };
+                updateResource(config.urls.api + '/flow/process-groups/' + encodeURIComponent(nfCanvasUtils.getGroupId()) + '/controller-services', entity).done(updateProcessGroup);
+            } else {
+                var componentsToStop = selection.filter(function (d) {
+                    return nfCanvasUtils.isProcessGroup(d3.select(this));
+                });
+
+                // ensure there are some component to stop
+                if (!componentsToStop.empty()) {
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:923]: Same comment as above regarding the verb used here.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 7]
```diff
@@ -845,6 +845,116 @@
...
+                        // prepare the request
+                        var uri, entity;
+                        uri = config.urls.api + '/flow/process-groups/' + encodeURIComponent(d.id) + '/controller-services';
+                        entity = {
+                            'id': d.id,
+                            'state': 'ENABLED'
+                        };
+
+                        startRequests.push(updateResource(uri, entity).done(function (response) {
+                            nfCanvasUtils.getComponentByType('ProcessGroup').reload(d.id);
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:884]: In the context of starting a Processor, this reload call was necessary to update the Process Group node on the canvas. There currently are no Controller Service statuses rendered on the Process Group node. I don't believe this call is necessary here.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js [hunk 8]
```diff
@@ -845,6 +845,116 @@
...
+                        // prepare the request
+                        var uri, entity;
+                        uri = config.urls.api + '/flow/process-groups/' + encodeURIComponent(d.id) + '/controller-services';
+                        entity = {
+                            'id': d.id,
+                            'state': 'DISABLED'
+                        };
+
+                        stopRequests.push(updateResource(uri, entity).done(function (response) {
+                            nfCanvasUtils.getComponentByType('ProcessGroup').reload(d.id);
```
Comment (mcgilman) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:939]: Same comment as above regarding the `reload` call.

**Review by mcgilman — CHANGES_REQUESTED** [review]
Thanks again for the PR @eduardofontes! I did comment previously regarding some important pre-conditions. However, based on @markap14 feedback, it sounds like this PR is still helpful as-is. I've reviewed the changeset and left some comments.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 1 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:854]: Callback function removed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 2 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:909]: Callback function removed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 3 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:863]: I've changed the condition to `if (!componentsToEnable.empty() && selection.size() === 1)`

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 4 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:918]: I've changed the condition to `if (!componentsToDisable.empty() && selection.size() === 1)`

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 5 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:868]: Done

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 6 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:923]: Done

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 7 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:884]: Actually it updates the processors status after enable/disable de controllers. I think it's better to keep this.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 8 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js:939]: Same comment above to keep this. :)

**Comment by eduardofontes** [comment]
> Thanks again for the PR @eduardofontes! I did comment previously regarding some important pre-conditions. However, based on @markap14 feedback, it sounds like this PR is still helpful as-is. I've reviewed the changeset and left some comments.

Thank you @mcgilman .
I've made the most of changes requested.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js (hunk 3 above)
Comment (eduardofontes) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-ui/src/main/webapp/js/nf/canvas/nf-actions.js]: @mcgilman I've refactored the functions and now they're simpler.

**Comment by mcgilman** [comment]
Thanks for the PR @eduardofontes! Looks great. This has been merged to main.

**Comment by andrewmlim** [comment]
Filed https://issues.apache.org/jira/browse/NIFI-9829 to capture some of the UX improvements discussed.

**Pull request merged**